Use &str in place of &[u8] where appropriate

matejcik opened the pull request:
The current Rust layout is not string-aware and passes bytes everywhere.

This PR converts most usage sites to &str, String and friends, so that we have UTF-8 enforced in the type system.

A `StrBuffer` type is introduced in the FFI to load `str` from MicroPython side properly.

grdddj reviewed (commented):
I am not sure what is wrong, but I have issues with `trezorctl set label hello` on this branch on emulator built with `UI2=1 make build_unix`. (It works fine on `master` build with `UI2=1`)

It is actually the same error I was getting in my storage branch when trying to use the `StrBuffer` instead of `Buffer`

```
495399016176 trezor.loop ERROR exception:
Traceback (most recent call last):
  File "trezor/loop.py", line 186, in _step
  File "apps/management/apply_settings.py", line 62, in apply_settings
  File "apps/management/apply_settings.py", line 119, in require_confirm_change_label
  File "trezor/ui/layouts/tt_v2/__init__.py", line 96, in confirm_action
TypeError: 
495399016241 trezor.workflow DEBUG close: <generator object 'apply_settings' at 7fb1afa2c060>
495399016272 trezor.workflow DEBUG default already started
495399016321 trezor.wire ERROR exception:
Traceback (most recent call last):
  File "trezor/wire/__init__.py", line 341, in _handle_single_message
  File "trezor/loop.py", line 533, in __iter__
  File "trezor/loop.py", line 186, in _step
  File "apps/management/apply_settings.py", line 62, in apply_settings
  File "apps/management/apply_settings.py", line 119, in require_confirm_change_label
  File "trezor/ui/layouts/tt_v2/__init__.py", line 96, in confirm_action
TypeError: 
```

matejcik commented:
should be fixed in [c62050f](https://github.com/trezor/trezor-firmware/pull/2209/commits/c62050f06ea9ff6b1c7a0f88f2f0b71fbc1e57f2)

grdddj reviewed (approved):
> should be fixed in [c62050f](https://github.com/trezor/trezor-firmware/pull/2209/commits/c62050f06ea9ff6b1c7a0f88f2f0b71fbc1e57f2)

Works for me. I do not have any other objections/suggestions

matejcik commented on core/embed/rust/src/micropython/buffer.rs:
```diff
@@ -132,6 +132,50 @@ impl AsMut<[u8]> for BufferMut {
+impl AsRef<str> for StrBuffer {
+    fn as_ref(&self) -> &str {
+        // SAFETY: MicroPython ensures that values of type `str` are UTF-8
+        unsafe { str::from_utf8_unchecked(self.0.as_ref()) }
```
that's a great question, and I'm tempted to stick some sort of validation at time of object creation for this case.

any idea what's the stable supported equivalent of the linked code?

mmilata replied:
I think we can just call [str::from_utf8](https://doc.rust-lang.org/std/str/fn.from_utf8.html)?

matejcik replied:
i originally wanted to do the check at creation time but then opted for `from_utf8().unwrap()` here instead, [26eda5c](https://github.com/trezor/trezor-firmware/pull/2209/commits/26eda5c88850ffc4d5f6b667a793a7f7b0fc6c9c)

matejcik commented:
squashed the fixup in [93e96d0](https://github.com/trezor/trezor-firmware/pull/2209/commits/93e96d0b8e06828900e133ab0833192657ee3b32), added the from_uft8 change on top of that

matejcik commented:
one more rebase on master to hopefully fix the weird "ui changes" failure
